Homework 2 on SmartHome

DaniilVyatkin opened the pull request:
Sorry for the strange branch name

DaniilVyatkin committed:
Homework 1 on SmartHome

DaniilVyatkin committed:
Homework 2
Unfortunately, I was unable to correctly implement turning off all the lights on closing the hall door.

a-ivanov commented:
Не виду следы 2го дз. Перечитайте задание https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5-2

DaniilVyatkin commented:
Прошу прощения, не запушил из локальной ветки. Сейчас все исправил, должны быть видны изменения, касающиеся 2 дз.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SmartHome.java:
```diff
@@ -14,6 +14,14 @@ public SmartHome(Collection<Room> rooms) {
+
     public void addRoom(Room room) {
         rooms.add(room);
     }
```
getRooms надо удбарть по 2 дз.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Room.java:
```diff
@@ -13,6 +13,17 @@ public Room(Collection<Light> lights, Collection<Door> doors, String name) {
+        }
+    }
+
     public Collection<Light> getLights() {
```
Мы не даем доступ по 2 дз к внутренней структуре дома.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/CloseDoorAction.java:
```diff
@@ -0,0 +1,26 @@
+                System.out.println("Door " + id + " was closed.");
+            }
+        }
+        if (actionable instanceof Room) {
```
Из-за этого над дверью действие два раза совершится

DaniilVyatkin replied:
Действие над дверью совершалось два раза не из-за этого, а из-за того, что CloseDoorAction создавался и в HallDoorEventHandler, и в DoorEventHandler. Я это исправил.

DaniilVyatkin committed:
Homework 2 corrections
1. Убрал getRooms и addRooms из SmartHome
2. Убрал getLights из Room
3. Исправил создание Action в HallDoorEventHandler, чтобы закрытие двери не совершалось два раза

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/CloseHallDoorAction.java:
```diff
@@ -0,0 +1,30 @@
+                for (Door door: ((Room) actionable).getDoors()) {
+                    if (door.getId().equals(this.id)) {
+                        System.out.println("Команда на закрытие входной двери. " +
+                                "Нужно выключить весь свет, но я не знаю, как:(");
```
Можно передать smartHome полем в CloseHallDoorAction, без добавления в сигнатуру Action.
smartHome - это деталь реализации этого Action, поэтому через конструктор полем.

a-ivanov replied:
Далее создаете тут нужное действие и применяете к дому.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/CloseHallDoorAction.java:
```diff
@@ -0,0 +1,30 @@
+        // По плану нужно было, чтобы при закрытии входной двери вызывался CloseHallDoorAction от всего дома,
+        // и тогда везде выключался бы свет.
+        // Но я не придумал, как передать сюда smartHome без добавления его в сигнатуру интерфейса Action
+        if (actionable instanceof SmartHome) {
```
Это сейчас просто выключает свет во всем доме, надо выключать только при условии, которые у вас ниже (`if (actionable instanceof Room) { ... }`).

DaniilVyatkin committed:
Corrections for Homework-2
1. Добавил smartHome в конструктор CloseHallDoorAction
2. Добавил TurnAllLightsOffAction - вызывается из CloseHallDoorAction, если оказывается, что закрыли входную дверь
